flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.maintenance.operator;
+
+import java.util.Set;
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.io.BulkDeletionFailureException;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.SupportsBulkOperations;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Delete the files using the {@link FileIO} which implements {@link SupportsBulkOperations}. */
+@Internal
+public class DeleteFilesProcessor extends AbstractStreamOperator<Void>
+    implements OneInputStreamOperator<String, Void> {
+  private static final Logger LOG = LoggerFactory.getLogger(DeleteFilesProcessor.class);
+
+  private final String name;
+  private final SupportsBulkOperations io;
+  private final String tableName;
+  private final Set<String> filesToDelete = Sets.newHashSet();
+  private final int batchSize;
+
+  private transient Counter failedCounter;
+  private transient Counter succeededCounter;
+
+  public DeleteFilesProcessor(String name, TableLoader tableLoader, int batchSize) {
+    Preconditions.checkNotNull(name, "Name should no be null");
+    Preconditions.checkNotNull(tableLoader, "Table loader should no be null");
+
+    tableLoader.open();
```

Comment:
table loader seems not closed. I actually think we don't need a `TableLoader` here. A read-only `SerializableTable` should work too.

Reply:
These are internal classes.
The parameters which are needed are:
- FileIO - for deleting
- TableName - for logging

We can move the code part for getting these to the `ExpireSnapshots` class, but I think the separation of concerns a bit better in this way. `ExpireSnapshots` is responsible for the flow, `DeleteFilesProcessor` is responsible for getting the data.

Also, the tableLoader is actually coming from the `TableMaintenance.forTable`, `TableMaintenance.forChangeStream` method, so it is better not to close them. I open these mostly for testing purposes, just to make sure they are open when we start using them.

Reply:
It is difficult to track if open and close are not handled by the same class. `TableLoader.clone` could be used if there are multiple places where table loader need to be opened and loaded.

Again,  a read-only `SerializableTable` is sufficient here, right? Can we have the `TableMaintenance` load the table once and pass the read-only `SerializableTable` to the `MaintenanceTaskBuilder.append(...)`?

Reply:
For this operator the table would be enough, but the `ExpireSnapshots` operator needs a TableLoader, so it is not enough to pass the read-only SerializableTable to the MaintenanceTaskBuilder.append(...).

Calling TableLoader.clone every time when we pass it as a parameter seems excessive to me. Especially considering that we would close them only on job restart. Do we follow the same pattern with the other Flink operators?

Reply:
I meant `MaintenanceTaskBuilder.append(...)` can pass in both `TableLoader` and `SerializableTable`.

> Calling TableLoader.clone every time when we pass it as a parameter seems excessive to me.

I meant clone when the lifecycle management needs to be different.

here is an example from `ContinuousSplitPlannerImpl`
````
  public ContinuousSplitPlannerImpl(
      TableLoader tableLoader, ScanContext scanContext, String threadName) {
    this.tableLoader = tableLoader.clone();
    this.tableLoader.open();
    this.table = this.tableLoader.loadTable();
...
  }

  @Override
  public void close() throws IOException {
...
    tableLoader.close();
  }
```

another example from `IcebergSource`
```
  private List<IcebergSourceSplit> planSplitsForBatch(String threadName) {
...
    try (TableLoader loader = tableLoader.clone()) {
      loader.open();
...
  }
```

Reply:
Done